demos/qml_optimization.ipynb

```diff
@@ -0,0 +1,469 @@
+        "    params, opt_state, _, _ = update_step(i, (params, opt_state, data, targets))\n",
+        "    loss_val = loss_fn(params, data, targets)\n",
+        "\n",
+        "    if i % 5 == 0:\n",
+        "        print(f\"Step: {i} Loss: {loss_val}\")\n",
+        "\n",
+        "    loss_history.append(loss_val)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "99kM0UDs83QM"
+      },
+      "source": [
+        "## JIT-compiling the optimization\n",
+        "\n",
+        "In the above example, we just-in-time (JIT) compiled our cost function `loss_fn`. However, we can also JIT compile the entire optimization loop; this means that the for-loop around optimization is not happening in Python, but is compiled and executed native. This avoids (potentially costly) data transfer between Python and our JIT compiled cost function with each update step."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 41,
+      "metadata": {
+        "id": "nZjOeD-79zKT"
+      },
+      "outputs": [],
+      "source": [
+        "params = {\"weights\": weights, \"bias\": bias}\n",
+        "\n",
+        "@qml.qjit\n",
+        "def optimization(params, data, targets):\n",
+        "    opt_state = opt.init(params)\n",
+        "    args = (params, opt_state, data, targets)\n",
+        "    (params, opt_state, _, _) = catalyst.for_loop(0, 100, 1)(update_step)(args)\n",
+        "    return params"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "dsFXWti89HQ_"
+      },
+      "source": [
+        "Note that we use `catalyst.for_loop` rather than a standard Python for loop, to allow the control flow to be JIT compatible."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 42,
+      "metadata": {
+        "id": "gVRxMBgg2nGF"
+      },
+      "outputs": [],
+      "source": [
+        "final_params = optimization(params, data, targets)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 43,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "kSJlW1rx2tyR",
+        "outputId": "a5020f22-fcc8-4921-93cb-81fea1a711f2"
+      },
+      "outputs": [
+        {
+          "name": "stdout",
+          "output_type": "stream",
+          "text": [
+            "{'bias': array(-0.75292885), 'weights': array([[ 1.63086995,  1.55018972,  0.6721261 ],\n",
+            "       [ 0.7266062 ,  0.36422543, -0.756247  ],\n",
+            "       [ 2.78387487,  0.62721014,  3.44996393],\n",
+            "       [-1.10119515, -0.12679488,  0.89283774],\n",
+            "       [ 1.27236329,  1.10631134,  2.22051434]])}\n"
+          ]
+        }
+      ],
+      "source": [
+        "print(final_params)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {},
+      "source": [
+        "## Timing the optimization"
```

Comment:
Should the demo show raw pennylane as well?

Reply:
Raw PL as in jax + default qubit, or raw PL as in lightning.qubit?

Interestingly:

- JAX + default.qubit is significantly faster, at 0.0031725019999839787 sec per loop
- lightning.qubit is significantly slower, at ~1 minute per loop.

I left out the lightning comparison because it was taking too long to time lol

Reply:
make sense! as jax + default.qubit use the JAX + XLA toolchain without any callbacks while lightning.qubit uses callbacks..

Reply:
@maliasadi What kind of callbacks are you referring to?

Reply:
I think Josh is just comparing "jax.jit + default.qubit" versus "lightning.qubit" (no jax).

Reply:
yep :) Since the classical processing was quite 'light' in this case, I wasn't sure if jax + lightning.qubit was worth the comparison?

Reply:
Agreed, I don't think it would do much.

Regarding the JAX result, it is really impressive how much computation can be optimized and fused once you have access to the simulator implementation!

But 1 min for `lightning.qubit` seems shocking: ~60x over the compiled version where the optimization loop is not even compiled? :O

To answer your first question, I was just thinking `default.qubit` with no jax (for "raw PennyLane").